pkg/virt-operator/resource/apply/rbac.go

```diff
@@ -263,22 +264,59 @@ func arePolicyRulesEqual(role1 runtime.Object, role2 runtime.Object) (equal bool
 
 		return true
 	}
+	changeExistingPolicyRulesByRequired := func(existing, required *[]rbacv1.PolicyRule) (changed bool) {
+		if !arePolicyRulesEqual(*existing, *required) {
+			*existing = *required
+			return true
+		}
+		return false
+	}
+	changeExistingSubjectsByRequired := func(existingSubjects, requiredSubjects *[]rbacv1.Subject) (changed bool) {
+		if len(*existingSubjects) != len(*requiredSubjects) {
+			*existingSubjects = *requiredSubjects
+			return true
+		}
+
+		for i := 0; i < len(*existingSubjects); i++ {
+			if (*existingSubjects)[i] != (*requiredSubjects)[i] {
+				*existingSubjects = *requiredSubjects
+				return true
+			}
+		}
+
+		return false
+	}
+	changeExistingRoleRefByRequired := func(existingRoleRef, requiredRoleRef *rbacv1.RoleRef) (changed bool) {
+		if *existingRoleRef != *requiredRoleRef {
+			*existingRoleRef = *requiredRoleRef
+			return true
+		}
+
+		return false
+	}
 
-	switch role1.(type) {
+	switch existing.(type) {
 	case *rbacv1.Role:
-		role1Obj := role1.(*rbacv1.Role)
-		role2Obj := role2.(*rbacv1.Role)
-		equal = arePolicyRulesEqualHelper(role1Obj.Rules, role2Obj.Rules)
+		existingRole := existing.(*rbacv1.Role)
+		requiredRole := required.(*rbacv1.Role)
+		changed = changeExistingPolicyRulesByRequired(&existingRole.Rules, &requiredRole.Rules)
 	case *rbacv1.ClusterRole:
-		role1Obj := role1.(*rbacv1.ClusterRole)
-		role2Obj := role2.(*rbacv1.ClusterRole)
-		equal = arePolicyRulesEqualHelper(role1Obj.Rules, role2Obj.Rules)
-	// Bindings do not have "rules" attribute
-	default:
-		equal = true
+		existingClusterRole := existing.(*rbacv1.ClusterRole)
+		requiredClusterRole := required.(*rbacv1.ClusterRole)
+		changed = changeExistingPolicyRulesByRequired(&existingClusterRole.Rules, &requiredClusterRole.Rules)
+	case *rbacv1.RoleBinding:
+		existingRoleBinding := existing.(*rbacv1.RoleBinding)
+		requiredRoleBinding := required.(*rbacv1.RoleBinding)
+		changed = changeExistingSubjectsByRequired(&existingRoleBinding.Subjects, &requiredRoleBinding.Subjects)
+		changed = changeExistingRoleRefByRequired(&existingRoleBinding.RoleRef, &requiredRoleBinding.RoleRef) || changed
```

Comment:
```
changed = changeExistingSubjectsByRequired(&existingRoleBinding.Subjects, &requiredRoleBinding.Subjects) || changeExistingRoleRefByRequired(&existingRoleBinding.RoleRef, &requiredRoleBinding.RoleRef)
```

or 

```
subjectsDifferent := changeExistingSubjectsByRequired(&existingRoleBinding.Subjects, &requiredRoleBinding.Subjects)
// short circuit if the first is true. 
changed = subjectsDifferent || changeExistingRoleRefByRequired(&existingRoleBinding.RoleRef, &requiredRoleBinding.RoleRef) 
```

also applies to the ClusterRoleBinding case

Reply:
The whole idea here is to avoid short circuit if the first is true.

Please note that `changeExistingSubjectsByRequired` and `changeExistingRoleRefByRequired` functions update the "existing" object by the required one. So if `changeExistingSubjectsByRequired` returned `true` it doesn't mean that I don't want to execute `changeExistingRoleRefByRequired` - I want to always execute both functions in order the perform all necessary updates. Only at the end I need to know if any modification occured.

But again, I see where the confusion is coming from.
I think it's best to break it into two variables: `subjectsUpdated` and `roleRefUpdated`, or something similar.

Reply:
Do we need to update the existing copy and update using it? could we instead just check if something is different and if it is update to the required spec? then we could short circuit as soon as something doesn't match

Reply:
I saw that this is how it's implemented in OpenShift, there they update `existing` by whatever necessary from `required`. My guess is that they did it this way to be on the safe side (not taking anything from `required` implicitly).

But you're right, there is a performance overhead for this approach. Not sure it's worth it.

What do you think?